File: bookstore/mern/controllers/review.controller.js
import { bookService, reviewService } from '../services/index.js'
import { errorHandler } from '../shared/helpers.js'

export default class ReviewController {
  async create(req, res, next) {
    return errorHandler(async () => {
      const { _book, author, content } = req.body
      const review = await reviewService.create(author, content)
      await bookService.addReview(_book, review._id)
      return res.json(review)
    }, next)
  }

  async change(req, res, next) {
    return errorHandler(async () => {
      const review = await reviewService.change(req.body._id, req.body.content)
      return res.json(review)
    }, next)
  }

  async delete(req, res, next) {
    return errorHandler(async () => {
      const id = await reviewService.delete(req.body._id)
      await bookService.removeReview(id)
      return res.json(id)
    }, next)
  }
}

File: bookstore/spring-backend/src/main/kotlin/masha/bookstore/controllers/ReviewController.kt
package masha.bookstore.controllers

import masha.bookstore.dtos.ReviewDto
import masha.bookstore.interfaces.IBookService
import masha.bookstore.interfaces.IReviewService
import masha.bookstore.models.Review
import masha.bookstore.shared.Route
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(Route.REVIEW)
@CrossOrigin(originPatterns = ["*"])
class ReviewController(
    private val reviewService: IReviewService,
    private val bookService: IBookService
) {

    @PostMapping
    fun create(@RequestBody dto: ReviewDto): ReviewDto {
        val review = reviewService.create(
            book = bookService.getOne(dto.book),
            author = dto.author,
            content = dto.content
        )

        bookService.addReview(review)
        return review.toDto()
    }

    @PatchMapping
    fun change(@RequestBody dto: ReviewDto): ReviewDto {
        return reviewService.change(id = dto.id, content = dto.content).toDto()
    }

    @DeleteMapping
    fun delete(@RequestBody dto: ReviewDto): Int {
        val review: Review = reviewService.getOne(dto.id)
        bookService.removeReview(review)
        return reviewService.delete(dto.id)
    }
}